web/pandas/pdeps/0013-standardize-apply.md

```diff
@@ -0,0 +1,151 @@
+| other callables (non ufunc)        | Try elements, fallback to series | series                           | series    |
+| list/dict of callables (non-ufunc) | Try elements, fallback to series | Try elements, fallback to series | series    |
+
+### apply
+
+|                                    | Series   | DataFrame   | groupby   |
+|:-----------------------------------|:---------|:------------|:----------|
+| ufunc or list/dict of ufuncs       | series   | series      | series    |
+| other callables (non ufunc)        | elements | series      | series    |
+| list/dict of callables (non-ufunc) | Try elements, fallback to series | series    | series    |
+
+The 3 tables show that:
+
+1. when given numpy ufuncs, callables given to `agg`/`transform`/`apply` operate on series data
+2. when used on groupby objects, callables given to `agg`/`transform`/`apply` operate on series data
+3. else, in some case it will try element-wise operation and fall back to series-wise operations if that fails, in some case will operate on series data and in some cases on element data.
+
+The above differences result on some non-obvious differences in how the same callable given to `agg`/`transform`/`apply` will behave.
+
+For example, calling `agg` using the same callable will give different results depending on context:
+
+```python
+>>> import pandas as pd
+>>> df = pd.DataFrame({"A": range(3)})
+>>>
+>>> df.agg(lambda x: np.sum(x))  # ok
+A    3
+dtype: int64
+>>> df.agg([lambda x: np.sum(x)])  # not ok
+         A
+  <lambda>
+0        0
+1        1
+2        2
+>>> df.A.agg(lambda x: np.sum(x))  # not ok
+0    0
+1    1
+2    2
+Name: A, dtype: int64
+```
+
+It can also have great effect on performance, even when the result is correct. For example:
+
+```python
+>>> df = pd.DataFrame({"A": range(1_000_000)})
+>>> %tiemit df.transform(lambda x: x + 1)  # fast
+1.43 ms ± 3.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
+ >>> %timeit df.transform([lambda x: x + 1])  # slow
+163 ms ± 754 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
+ >>> %timeit df.A.transform(lambda x: x + 1)  # slow
+162 ms ± 980 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
+```
+
+The reason for the great performance difference is that `df.transform(func)` operates on series data, which is fast, while `df.transform(func_list)` will attempt elementwise operation first, and if that works (which is does here), will be much slower than series operations.
+
+In addition to the above effects of the current implementation of `agg`/`transform` & `apply`, see [#52140](https://github.com/pandas-dev/pandas/issues/52140) for more examples of the unexpected effects of how `apply` is implemented.
+
+It can also be noted that `Series.apply` & `DataFrame.apply` could almost always be replaced with calls to `agg`, `transform` & `map`, if `agg` & `transform` were to always operate on series data. For some examples, see the table below for alternatives using `apply(func)`:
```

Comment:
```suggestion
It can also be noted that `Series.apply` & `DataFrame.apply` could almost always be replaced with calls to `agg`, `transform` or `map`, if `agg` and `transform` were to always operate on series data. For some examples, see the table below for alternatives using `apply(func)`:
```